rustls/src/stream.rs

```diff
@@ -62,24 +55,53 @@ where
             }
         }
 
+        Ok(())
+    }
+
+    // Implements `BufRead::fill_buf` but with more flexible lifetimes, so StreamOwned can reuse it
+    fn fill_buf(mut self) -> Result<&'a [u8]>
+    where
+        S: 'a,
+    {
+        self.prepare_read()?;
+        self.conn.reader().into_first_chunk()
+    }
+}
+
+impl<'a, C, T, S> Read for Stream<'a, C, T>
+where
+    C: 'a + DerefMut + Deref<Target = ConnectionCommon<S>>,
+    T: 'a + Read + Write,
+    S: SideData,
+{
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        self.prepare_read()?;
         self.conn.reader().read(buf)
     }
 
     #[cfg(read_buf)]
     fn read_buf(&mut self, cursor: core::io::BorrowedCursor<'_>) -> Result<()> {
-        self.complete_prior_io()?;
+        self.prepare_read()?;
+        self.conn.reader().read_buf(cursor)
+    }
+}
 
-        // We call complete_io() in a loop since a single call may read only
-        // a partial packet from the underlying transport. A full packet is
-        // needed to get more plaintext, which we must do if EOF has not been
-        // hit.
-        while self.conn.wants_read() {
-            if self.conn.complete_io(self.sock)?.0 == 0 {
-                break;
-            }
+impl<'a, C, T, S> BufRead for Stream<'a, C, T>
+where
+    C: 'a + DerefMut + Deref<Target = ConnectionCommon<S>>,
+    T: 'a + Read + Write,
+    S: 'a + SideData,
+{
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        Stream {
```

Comment:
Why do we need to reconstitute a `Stream` here?

Reply:
It reborrows, which is necessary since the helper takes `self` by value. I'll add a comment.